src/python/pants/backend/python/util_rules/pex_from_targets_test.py

```diff
@@ -759,6 +763,52 @@ def test_exclude_sources(include_sources: bool, rule_runner: PythonRuleRunner) -
     assert len(snapshot.files) == (1 if include_sources else 0)
 
 
+def test_include_sources_without_transitive_package_sources(rule_runner: PythonRuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "src/app/BUILD": dedent(
+                """
+                python_sources(
+                    name="app",
+                    sources=["app.py"],
+                    dependencies=["//src/dep:pkg"],
+                )
+                """
+            ),
+            "src/app/app.py": "",
+            "src/dep/BUILD": dedent(
+                # This test requires a package that has a standard dependencies field.
+                # 'pex_binary' has a dependencies field; 'archive' does not.
+                """
+                pex_binary(name="pkg", dependencies=[":dep"])
+                python_sources(name="dep", sources=["dep.py"])
+                """
+            ),
+            "src/dep/dep.py": "",
+        }
+    )
+
+    rule_runner.set_options(
+        [
+            "--backend-packages=pants.backend.python",
+            "--python-repos-indexes=[]",
+        ],
+        env_inherit={"PATH"},
+    )
+
+    request = PexFromTargetsRequest(
+        [Address("src/app", target_name="app")],
+        output_filename="demo.pex",
+        internal_only=True,
+        include_source_files=True,
+    )
+    pex_request = rule_runner.request(PexRequest, [request])
+    snapshot = rule_runner.request(Snapshot, [pex_request.sources])
+
+    # the packaged transitive dep is excluded
+    assert snapshot.files == ("app/app.py",)
```

Comment:
So... actually we might want to do what I suggested above and save this change for a different PR.

In this case I'd expect the packaged thing to be in the snapshot. The user explicitly declared a dependency from their app to a `pex_binary`. It's extremely rude to just ignore that :sweat_smile: . What is most likely is they plan on calling that `pex_binary` in their code. We shouldn't silently be ignoring the dep.

Otherwise this is a bit of a regression because now the user can't do _anything_ with that dependency, which they _explicitly_ provided.

Reply:
Not quite. This snapshot is for the source files that should be included in THIS pex (the pex requested with `PexFromTargetsRequest`). So, if you request multiple pexes that you intend to use together, including the same files can cause conflicts. So, whatever makes the `PexFromTargetsRequest` needs to handle including multiple packages, if needed, in the process sandbox (again: this snapshot is NOT the final process sandbox; this snapshot is for the sources in a single PEX when `include_sources=True`).

Reply:
It's still a change in behavior, so I don't feel confident approving this PR until we collectively agree with the change in behavior (which will almost definitely involve some kind of warning/deprecation).

And because we don't know whether the user intends to embed this PEX in the other one (e.g. `resource`) or just expect it exists at runtime (e.g. `file`) there really isn't a one, true answer anyways.

Reply:
If I'm not mistaken, this PR does not change the presence of the depended upon pex, but its sources, so

> And because we don't know whether the user intends to embed this PEX in the other one (e.g. resource) or just expect it exists at runtime (e.g. file) there really isn't a one, true answer anyways.

is unaffected here. It's whether `src/dep/dep.py` is present or not for the `demo.pex` sandbox. I agree with the notion that as there is _not_ a dependency from `src/app/app.py` to `src/dep/dep.py` it is correct to not include it in the sandbox.

Currently it is, as it is a transitive dependency, however it doesn't make sense to traverse through to a packaged targets sources. That's like including the cake _and_ all it's ingredients as well.

Reply:
Right, we can all agree the current behavior is wrong. But what the expected behavior should be, and how to handle the change in behavior is something we can and should address separately.

Right now this PR doesn't honor our deprecation policy, since it is actually a User API Change.